File: bbb/bbb/report/advance_booking_list_report/advance_booking_list_report.js
// Copyright (c) 2023, invento software limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Advance Booking List Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
		},
		
		{
			"fieldname": "status",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": ["", "To Deliver and Bill", "To Bill", "To Deliver", "Completed"],
			"width": "60px"
		},
		{
			"fieldname": "billing_status",
			"label": __("Billing Status"),
			"fieldtype": "Select",
			"options": ["", "Not Billed", "Fully Billed", "Partly Billed", "Closed"],
			"width": "60px"
		},
		{
			"fieldname": "outlet",
			"label": __("Outlet"),
			"fieldtype": "Link",
			"options": "POS Profile",
			"width": "60px"
		},
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_default('company'),
			// "reqd": 1,
			"width": "60px"
		},
	],
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname == "status" && data.status === "Completed"){
			value = "<span style='color:green'>" + value + "</span>";
		  }
		else if(column.fieldname == "status"){
			value = "<span style='color:dark'>" + value + "</span>";
		}
		if (column.fieldname == "billing_status" && data && data.billing_status === "Not Billed"){
		  value = "<span style='color:red'>" + value + "</span>";
		}
		else if (column.fieldname == "billing_status" && data && data.billing_status === "Fully Billed"){
			value = "<span style='color:green'>" + value + "</span>";
		  }
		else if (column.fieldname == "billing_status" && data && data.billing_status === "Partly Billed"){
			value = "<span style='color:orange'>" + value + "</span>";
		}
		else if (column.fieldname == "billing_status" && data && data.billing_status === "Closed"){
			value = "<span style='color:dark'>" + value + "</span>";
		}
		return value;
	  },
};

File: bbb/bbb/report/advance_booking_list_report/advance_booking_list_report.py
# Copyright (c) 2022, invento software limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _
import copy


def execute(filters=None):
    columns, data = get_columns(), get_invoice_data(filters)
    return columns, data


def get_columns():
    """return columns"""
    columns = [
        # {"label": _("No"), "fieldname": "serial_no", "fieldtype": "Text", "width": 100},
        {"label": _("Booking Id"), "fieldname": "name", "fieldtype": "Link", "options": "Advance Booking",
         "width": 180},
        {"label": _("Outlet"), "fieldname": "pos_profile", "fieldtype": "Link", "options": "POS Profile",
         "width": 150},
        {"label": _("Posting Date"), "fieldname": "posting_date",
         "fieldtype": "Date", "width": 110},
        {"label": _("Service Date"), "fieldname": "actual_service_date",
         "fieldtype": "Date", "width": 110},
        {"label": _("Status"), "fieldname": "status",
         "fieldtype": "Data", "width": 130},
        {"label": _("Billing Status"), "fieldname": "billing_status",
         "fieldtype": "Data", "width": 110},
        {"label": _("Total Qty"), "fieldname": "total_qty",
         "fieldtype": "Int", "width": 85},
        {"label": _("Total"), "fieldname": "total", "fieldtype": "Currency", "width": 100,
         "convertible": "rate", "options": "currency"},
        {"label": _("Discount"), "fieldname": "discount_amount", "fieldtype": "Currency", "width": 100,
         "convertible": "rate", "options": "currency"},
        {"label": _("Grand Total"), "fieldname": "grand_total", "fieldtype": "Currency", "width": 110,
         "convertible": "rate", "options": "currency"},
        {"label": _("Rounded Total"), "fieldname": "rounded_total", "fieldtype": "Currency", "width": 115,
         "convertible": "rate", "options": "currency"},
        {"label": _("Total Advance"), "fieldname": "total_advance", "fieldtype": "Currency", "width": 115,
         "convertible": "rate", "options": "currency"},
        {"label": _("Total Due"), "fieldname": "outstanding_amount", "fieldtype": "Currency", "width": 110,
         "convertible": "rate", "options": "currency"},
    ]
    return columns


def get_conditions(filters):
    conditions = []

    if filters.get("from_date"):
        conditions.append("advance_booking.actual_service_date >= '%s'" %
                          filters.get("from_date"))

    if filters.get("to_date"):
        conditions.append("advance_booking.actual_service_date <= '%s'" %
                          filters.get("to_date"))
        
    if filters.get("billing_status"):
        conditions.append("advance_booking.billing_status = '%s'" %
                          filters.get("billing_status"))
        
    if filters.get("status"):
        conditions.append("advance_booking.status = '%s'" %
                          filters.get("status"))

    if filters.get("outlet"):
        conditions.append("advance_booking.pos_profile = '%s'" %filters.get("outlet"))
        
    if filters.get("company"):
        conditions.append("advance_booking.company = '%s'" %
                          filters.get("company"))

    if conditions:
        conditions = " and ".join(conditions)
    return conditions


def get_invoice_data(filters):
    conditions = get_conditions(filters)
    query_result = frappe.db.sql(""" SELECT advance_booking.pos_profile,
       	advance_booking.name,
    	advance_booking.pos_profile,
    	advance_booking.posting_date,
    	advance_booking.actual_service_date,
    	advance_booking.total_qty,
    	advance_booking.billing_status,
    	advance_booking.status,
    	advance_booking.total,
    	advance_booking.discount_amount,
     	advance_booking.grand_total,
     	advance_booking.rounded_total,
     	advance_booking.total_advance,
     	advance_booking.outstanding_amount
      	FROM `tabAdvance Booking` advance_booking where advance_booking.docstatus = 1 and %s order by advance_booking.name""" % (conditions), as_dict=1)

    return query_result
